src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -1547,7 +1548,11 @@ private void HandleOutOfMemoryException(Exception exception, IndexingStatsScope
                 if (_logger.IsInfoEnabled)
                     _logger.Info($"Out of memory occurred for '{Name}'", exception);
 
-                DocumentDatabase.NotificationCenter.OutOfMemory.Add(_environment, exception);
+                //We are in low of memory so we want to allocate the strings only once
+                if (_environmentStr == null)
```

Comment:
You haven't changed much. You are still allocating _when you are in OOM condition_

Reply:
on the other hand, the for no InMemory databases, .ToString() eventually return a reference to the config path value, that is already materialized, so I don't think it's a big issue here.
But maybe we should return to using the Environment, but use a ConditionalTable instead of ConcurrentDictionary

Reply:
I think that this is a much better option, yes.